core/src/main/java/org/elasticsearch/cluster/service/TaskBatching.java

```diff
@@ -0,0 +1,233 @@
+        if (tasks.isEmpty()) {
+            return;
+        }
+        final BatchingTask firstTask = tasks.get(0);
+        assert tasks.stream().allMatch(t -> t.batchingKey == firstTask.batchingKey) :
+            "tasks submitted in a batch should share the same batching key: " + tasks;
+        // convert to an identity map to check for dups based on task identity
+        final Map<Object, BatchingTask> tasksIdentity = tasks.stream().collect(Collectors.toMap(
+            BatchingTask::getTaskIdentity,
+            Function.identity(),
+            (a, b) -> { throw new IllegalStateException("cannot add duplicate task: " + a); },
+            IdentityHashMap::new));
+
+        synchronized (tasksPerExecutor) {
+            LinkedHashSet<BatchingTask> existingTasks = tasksPerExecutor.computeIfAbsent(firstTask.batchingKey,
+                k -> new LinkedHashSet<>(tasks.size()));
+            for (BatchingTask existing : existingTasks) {
+                // check that there won't be two tasks with the same identity for the same batching key
+                BatchingTask duplicateTask = tasksIdentity.get(existing.getTaskIdentity());
+                if (duplicateTask != null) {
+                    throw new IllegalStateException("task [" + duplicateTask.describeTasks(
+                        Collections.singletonList(existing)) + "] with source [" + duplicateTask.source + "] is already queued");
+                }
+            }
+            existingTasks.addAll(tasks);
+        }
+
+        if (timeout != null) {
+            threadExecutor.execute(firstTask, threadPool.scheduler(), timeout, () -> onTimeoutInternal(tasks, timeout));
+        } else {
+            threadExecutor.execute(firstTask);
+        }
+    }
+
+    private void onTimeoutInternal(List<? extends BatchingTask> updateTasks, TimeValue timeout) {
+        threadPool.generic().execute(() -> {
+            final ArrayList<BatchingTask> toRemove = new ArrayList<>();
+            for (BatchingTask task : updateTasks) {
+                if (task.processed.getAndSet(true) == false) {
+                    logger.debug("task [{}] timed out after [{}]", task.source, timeout);
+                    toRemove.add(task);
+                }
+            }
+            if (toRemove.isEmpty() == false) {
+                Object batchingExecutor = toRemove.get(0).batchingKey;
+                synchronized (tasksPerExecutor) {
+                    LinkedHashSet<BatchingTask> existingTasks = tasksPerExecutor.get(batchingExecutor);
+                    if (existingTasks != null) {
+                        existingTasks.removeAll(toRemove);
+                        if (existingTasks.isEmpty()) {
+                            tasksPerExecutor.remove(batchingExecutor);
+                        }
+                    }
+                }
+                for (BatchingTask task : toRemove) {
```

Comment:
can we maybe change onTimeout(task, timeout) to be onTimeout(List<..> tasks, timeout, batchingKey) ? than we can leave the spawning to the implementation. This means we don't need to capture a thread pool for this class. Also I'm wondering if we should fold `threadPool.scheduler()` into `PrioritizedEsThreadPoolExecutor` as I don't see the need for people to use multiple schedulers.